extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js

```diff
@@ -325,15 +330,52 @@ function sendAmpScriptToSwOnFirstVisit(registration, win) {
           .filter(item => item.initiatorType === 'script' &&
             item.name.indexOf(urls.cdn) !== -1)
           .map(script => script.name);
-      // using https://github.com/redux-utilities/flux-standard-action
+      // using convention from https://github.com/redux-utilities/flux-standard-action.
       controllerSw.postMessage(JSON.stringify(dict({
-        'type': 'FIRST_VISIT_CACHING',
+        'type': 'AMP__FIRST-VISIT-CACHING',
         'payload': ampScriptsUsed,
       })));
     });
   }
 }
 
+/**
+ * Whenever a new service worker is activated, controlled page will send
+ * the used AMP scripts and the self's URL to service worker to be cached.
+ * @param {ServiceWorkerRegistration} registration
+ * @param {!Window} win
+ */
+function prefetchOutgoingLinks(registration, win) {
+  const {document} = win;
+  const links = [].map.call(document.querySelectorAll('a[data-rel=prefetch]'),
+      link => link.getAttribute('href'));
+  if (supportsPrefetch(document)) {
+    links.forEach(link => {
+      const linkTag = document.createElement('link');
+      linkTag.setAttribute('rel', 'prefetch');
+      linkTag.setAttribute('href', link);
+      document.head.appendChild(linkTag);
+    });
+  } else if (registration.active) {
+    registration.active.postMessage(JSON.stringify(dict({
+      'type': 'AMP__LINK-PREFETCH',
+      'payload': links,
+    })));
+  }
+}
+
+/**
+ * Returns whether or not link rel=prefetch is supported.
+ * @param {!Document} doc
+ * @return {boolean}
+ */
+function supportsPrefetch(doc) {
```

Comment:
I think we already have this check in preconnect?

Reply:
not prefetch, just pre-connect checks: https://github.com/ampproject/amphtml/blob/5f75efe35b734a4fcf0884d373315de60cd42df9/src/preconnect.js#L52-L67